lib/puppet/util/log.rb

```diff
@@ -170,6 +170,17 @@ def Log.with_destination(destination, &block)
   def Log.newmessage(msg)
     return if @levels.index(msg.level) < @loglevel
 
+    if msg.message.valid_encoding?
+      message = Puppet::Util::CharacterEncoding.convert_to_utf_8(msg.message)
+    else
+      message = _("Received a log message with invalid encoding:")
+      message << Puppet::Util::CharacterEncoding.convert_to_utf_8(msg.message.dump) << "\n"
+      # We only select the last 5 callers in the stack to avoid being spammy
```

Comment:
Sure. Maybe 10 lines? I really wouldn't want to go deeper than 10.. Also just realized definitely need to make sure we don't access out of bounds of the array. Classic.